File: BookingRequestTypes.scala
case class BookingRequest private(
  val date: Date,
  val seats: Seats
)

sealed trait Error

// These wrap errors from other modules.
case class DateError(e: Date.Error) extends Error
case class SeatsError(e: Seats.Error) extends Error

// Our additional errors.
case class DateBefore(date1: Date, date2: Date) extends Error
case class Missing(label: String) extends Error

File: BookingRequestTypes.hs
data BookingRequest =
  BookingRequest { getDate :: Date.Date
                 , getSeats :: Seats.Seats
                 }

data Error =
    DateError Date.Error
  | SeatsError Seats.Error
  | Missing String       -- ^ label
  | DateBefore Date.Date -- ^ date that was attempted
               Date.Date -- ^ the current date at attempt
